### ISSUE TITLE
URLs cannot contain a colon

### ISSUE THREAD

**Issue by fdw** [spring-projects/spring-hateoas#477]
I'm trying to create a resource whose ID contains a colon (f.e. "AB:TRE"). When I use createResourceWithId(), a UriComponentsBuilder is used to split the id wrongly into a scheme and ssp, even though it should only be part of the path. Thus, the resulting uri does not contain the id.

According to the relevant RFC 3986, colons are allowed in segments: 

> segment       = *pchar
> pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"

**Comment by fdw** [comment]
Is there some progress with this issue? Do you need more information?

**Comment by fdw** [comment]
This is a really annoying bug for us. Do you have any idea what we should do (besides "don't use colons")?

**Comment by gregturn** [comment]
@fdw Sorry for being so slow. Think we can draft a simple test case/class to reproduce this? I'd love to work on the patch for it.

**Comment by fdw** [comment]
Hello @gregturn, thank you for replying!

Regarding a test case, I think it should be enough to call `createResourceWithId` with a first parameter that contains a colon. For example:
```
createResourceWithId("abc:def", new Object());
```


I would expect the resulting serialized link to end with "url_for_controller/abc:def"; but the actual result does not contain any id at all (i.e. "url_for_controller/").

Is this enough, or do you need a more fleshed out example?

**Comment by gregturn** [comment]
@fdw I'll have to look at that in one week. Spring break starts tomorrow so I'll be offline at least for a week. But we are getting things moving on this project, if you haven't noticed already.

**Comment by fdw** [comment]
Thank you, that is good to hear! Enjoy your vacation!

**Comment by gregturn** [comment]
@fdw Can you check out if #582 fixes the problem you're seeing?

**Comment by rolandKulcsar** [comment]
@fdw @gregturn I wrote a test case which is green on me with the current master:
```java
@Test
public void usesAlternateIdWithColonIfGivenExplicitly() {

    PersonResource resource = assembler.createResourceWithId("abc:def", person);
    Link selfLink = resource.getId();
    assertThat(selfLink.getHref(), endsWith("/people/abc:def"));
}
```

**Comment by TingShingIp** [comment]
Issue should be solved with version 0.24

**Comment by gregturn** [comment]
Resolved via https://github.com/spring-projects/spring-hateoas/commit/d68700231eb1d8afaf27328cc5d576a71e966a2a.
